util/edns0.go

```diff
@@ -0,0 +1,130 @@
+package util
+
+import (
+	"slices"
+
+	"github.com/miekg/dns"
+)
+
+// EDNS0Option is an interface for all EDNS0 options as type constraint for generics.
+type EDNS0Option interface {
+	*dns.EDNS0_SUBNET | *dns.EDNS0_EDE | *dns.EDNS0_LOCAL | *dns.EDNS0_NSID | *dns.EDNS0_COOKIE | *dns.EDNS0_UL
+	Option() uint16
+}
+
+// RemoveEdns0Record removes the OPT record from the Extra section of the given message.
+func RemoveEdns0Record(msg *dns.Msg) bool {
+	if msg == nil || msg.IsEdns0() == nil || len(msg.Extra) == 0 {
+		return false
+	}
+
+	for i, rr := range msg.Extra {
+		if rr.Header().Rrtype == dns.TypeOPT {
+			msg.Extra = slices.Delete(msg.Extra, i, i+1)
+
+			return true
+		}
+	}
+
+	return false
+}
+
+// GetEdns0Record returns the OPT record from the Extra section of the given message.
+func GetEdns0Record(msg *dns.Msg) *dns.OPT {
+	if msg == nil {
+		return nil
+	}
+
+	if res := msg.IsEdns0(); res != nil {
+		return res
+	}
+
+	res := new(dns.OPT)
+	res.Hdr.Name = "."
+	res.Hdr.Rrtype = dns.TypeOPT
+	msg.Extra = append(msg.Extra, res)
+
+	return res
+}
+
+// GetEdns0Option returns the option with the given code from the OPT record in the
+// Extra section of the given message.
+func GetEdns0Option[T EDNS0Option](msg *dns.Msg) T {
+	code := getCode[T](msg)
+	if code == 0 {
+		return nil
+	}
+
+	opt := GetEdns0Record(msg)
+	for _, o := range opt.Option {
+		if o.Option() == code {
+			return o.(T)
```

Comment:
I guess it's not supposed to happens because `dns` wouldn't create a broken struct, but having a nice error message can't hurt:
```suggestion
			t, ok := o.(T)
			if !ok {
				panic(fmt.Errorf("dns option with code %d is not of type %T", code, t))
			}
```